## Instituto Federal de Minas Gerais campus Ibirité

<br>

**Curso**: Engenharia de controle e automação.

**Professor**: Ivan Reinaldo Meneguini.

**Disciplina**: Cálculo Numérico.

**Aula/Tema**: Aula 02-Método da bisseção.

<br>

*Mateus Menezes Carvalho, 10 de Março de 2023.*

<br>

1. Dê um exemplo de uma função contínua que tenha uma raiz no intervalo $[a, b]$ mas $f(a) \ast f(b) > 0$. Isto contraria o teorema do valor intermediário?

<br>

$$f(x) = x²$$

<br>

Sim. Existem alguns casos de funções contínuas, que em um determinado intervalo $[a, b]$, é possível determinar suas raizes quando $f(a) *(b) > 0$, é o caso da função $y = x²$.

<br>
2. Implemente o método da bisseção.
<br>

In [81]:
def bissecao(a, b, tol, funcao, intermax=100):


    """
    
    Retorna uma solução aproximada para p caso exista.

    """

    xa = a
    xb = b

    i = 1
    p = 0

    valor_a_f = funcao(a) # Valor de a avaliado em f(x)
    valor_p_f = funcao(p) # Valor de p avaliado em f(x)

    while i <= intermax:
        
        
        p = (a + b) / 2
        valor_p_f = funcao(p)

        if abs(valor_p_f) < tol or (b - a)/2 < tol:
            return f'f(x) possui raíz no intervalo ({xa}, {xb}). r = {p:.2f}, é uma aproximação para a raiz.'
        
        elif valor_a_f * valor_p_f > 0:
            a = p
            valor_p_f = valor_a_f

        else:
            b = p

        i=+1    

    return 'O método da bisseção falhou!'    

<br>

3. Justifique que a equação $4x-e^x = 0$ tem uma raiz no intervalo $(0, 1)$ e outra no intervalo $(2,3)$. Utilize o método da bisseção para encontrar uma aproximação para estas raízes. Use $tol = 10^⁻5$ e $itermax = 100$.

<br>

In [69]:
import math

# Função:
funcao = lambda x: (4*x) - (math.e**x)

# Aplicando o método da bisseção para o intervalo (0,1):
print(bissecao(0, 1, 10**-5, funcao))

# Aplicando o método da bisseção para o intervalo (2,3):
print(bissecao(2, 3, 10**-5, funcao))

f(x) possui raíz no intervalo (0, 1). r = 0.36, é uma aproximação para a raíz.
f(x) possui raíz no intervalo (2, 3). r = 2.15, é uma aproximação para a raíz.



Sendo $r$ o valor da raiz da equação $4x-e^x = 0$ no ponto $(x', x'')$: 

+ $f(x)$ possui raiz no intervalo $(0, 1)$. $r = 0.36$, é uma aproximação para a raiz.

+ $f(x)$ possui raiz no intervalo $(2, 3)$. $r = 2.15$, é uma aproximação para a raiz.

Portanto, quando $x = r$, $f(x) = 0$.


<br>

4. Utilize o método da bisseção para encontrar uma raiz das equações $f(x) = 0$ definidas pelas funções a seguir 
nos intervalos indicados. Use $tol = 10^⁻5$ e $itermax = 100$.

<br>

<br>

a) $$f(x) = \sqrt{x} - cos(x), \space I = [0, 1]$$

<br>

In [76]:
# Função:
funcao = lambda x: (x)**1/2 - (math.cos(x))

# Aplicando o método da bisseção para o intervalo (0,1):
print(bissecao(0, 1, 10**-5, funcao))

f(x) possui raíz no intervalo (0, 1). r = 1.000, é uma aproximação para a raiz.


$f(x)$ possui raíz no intervalo $(0, 1)$. $r = 1.00$, é uma aproximação para a raiz.

<br>

b) $$f(x) = 3(x + 1)(x - 0.5)(x - 1), \space I = [0, 1]$$

<br>

In [82]:
# Função:
funcao = lambda x: 3*(x + 1)*(x - 0.5)*(x - 1)

# Aplicando o método da bisseção para o intervalo (0,1):
print(bissecao(-3, 0, 10**-5, funcao))

f(x) possui raíz no intervalo (-3, 0). r = -1.00, é uma aproximação para a raiz.


f(x) possui raíz no intervalo (-3, 0). r = -1.00, é uma aproximação para a raiz.

<br>

b) $$f(x) = 3(x + 1)(x - 0.5)(x - 1), \space I = [0, 0.8]$$

<br>

In [83]:
# Função:
funcao = lambda x: 3*(x + 1)*(x - 0.5)*(x - 1)

# Aplicando o método da bisseção para o intervalo (0,1):
print(bissecao(0, 0.8, 10**-5, funcao))

f(x) possui raíz no intervalo (0, 0.8). r = 0.50, é uma aproximação para a raiz.


f(x) possui raíz no intervalo (0, 0.8). r = 0.50, é uma aproximação para a raiz.

<br>

c) $$f(x) = x - e^{-x}, \space I = [0, 1]$$

<br>

In [85]:
# Função:
funcao = lambda x: x - (math.e**-x)

# Aplicando o método da bisseção para o intervalo (0,1):
print(bissecao(0, 1, 10**-5, funcao))

f(x) possui raíz no intervalo (0, 1). r = 0.57, é uma aproximação para a raiz.


f(x) possui raíz no intervalo (0, 1). r = 0.57, é uma aproximação para a raiz.

<br>

e) $$f(x) = e^{x} - x² + 3x - 2, \space I = [0, 1]$$

<br>

In [86]:
# Função:
funcao = lambda x: (math.e**x) - (x**2) + (3*x) - 2

# Aplicando o método da bisseção para o intervalo (0,1):
print(bissecao(0, 1, 10**-5, funcao))

f(x) possui raíz no intervalo (0, 1). r = 0.26, é uma aproximação para a raiz.


f(x) possui raíz no intervalo (0, 1). r = 0.26, é uma aproximação para a raiz.

<br>

d) $$f(x) = x + 3 - sen(\pi x), \space I = [-1, 0]$$

<br>

In [89]:
# Função:
funcao = lambda x: (x + 3) - math.sin(math.pi*x)

# Aplicando o método da bisseção para o intervalo (-1, 0):
print(bissecao(-1, 0, 10**-5, funcao))

f(x) possui raíz no intervalo (-1, 0). r = -0.00, é uma aproximação para a raiz.


f(x) possui raíz no intervalo (-1, 0). r = -0.00, é uma aproximação para a raiz.

<br>

e) $$f(x) = x + 3 - sen(\pi x), \space I = [0, 0.5]$$

<br>

In [90]:
# Função:
funcao = lambda x: (x + 3) - math.sin(math.pi*x)

# Aplicando o método da bisseção para o intervalo (-1, 0):
print(bissecao(0, 0.5, 10**-5, funcao))

f(x) possui raíz no intervalo (0, 0.5). r = 0.50, é uma aproximação para a raiz.


f(x) possui raíz no intervalo (0, 0.5). r = 0.50, é uma aproximação para a raiz.

5. Determine uma aproximação para $\sqrt[3]{25}$ correta até $10^{⁻4}$.